### PR TITLE
[Java FFI][Bug 18846] Implement a parser for the Java FFI DSL

### PR DESCRIPTION
**Pull request by livecodeali**
Replaces https://github.com/livecode/livecode/pull/4438

### CODE REVIEW

File: toolchain/java-dsl-parse/src/LAYOUT.b [hunk 1]
```diff
@@ -0,0 +1,2 @@
+\  { AdvanceCurrentPosition(1); }
+\t {  }
```
Comment (peter-b) [inline toolchain/java-dsl-parse/src/LAYOUT.b]: You need to `AdvanceCurrentPosition(1);` for tab characters, or column numbers will be reported incorrectly when errors occur on lines indented with tabs.

File: toolchain/java-dsl-parse/src/DOUBLE_LITERAL.t [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[0-9]+"."[0-9]+([eE][-+]?[0-9]+)? {
```
Comment (peter-b) [inline toolchain/java-dsl-parse/src/DOUBLE_LITERAL.t]: It looks like `1e1` is not parsed as a numeric literal, although it is a numeric literal in LCB.  This isn't necessarily a problem, but it's probably something to be aware of.

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 4468582

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 4468582e5f1b27b87644b8ae4839c35b087f5997

**Comment by livecode-vulcan** [comment]
😞 test failure 4468582e5f1b27b87644b8ae4839c35b087f5997

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/998)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1004)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/996)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/997)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1004)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1018)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/995)

**Review by peter-b — APPROVED** [review]

File: toolchain/lc-compile-ffi-java/src/LAYOUT.b [hunk 3]
```diff
@@ -0,0 +1,2 @@
+\  { AdvanceCurrentPosition(1); }
+\t {  }
```
Comment (peter-b) [inline toolchain/lc-compile-ffi-java/src/LAYOUT.b]: You need to `AdvanceCurrentPosition` for tab characters, or the incorrect column number will be reported in error and warning messages.

**Review by peter-b — CHANGES_REQUESTED** [review]
Oops, I found one minor issue that I didn't flag up properly earlier.

File: toolchain/lc-compile-ffi-java/src/LAYOUT.b (hunk 3 above)
Comment (livecodeali) [reply toolchain/lc-compile-ffi-java/src/LAYOUT.b]: Good catch :-)

**Review by peter-b — APPROVED** [review]

**Comment by peter-b** [comment]
@livecode-vulcan review ok 5e452a7

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 5e452a723a2d9b7856c99b09bfb4bae7b7a27035

**Comment by livecode-vulcan** [comment]
😞 test failure 5e452a723a2d9b7856c99b09bfb4bae7b7a27035

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1003)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1009)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1001)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1002)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1009)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1023)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1000)

**Comment by peter-b** [comment]
@livecode-vulcan review ok 33f63c2

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 33f63c2eecfb71fde4d30e59d9803a90948ec522

**Comment by livecode-vulcan** [comment]
😎 test success 33f63c2eecfb71fde4d30e59d9803a90948ec522

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1004)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1010)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1002)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1003)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1010)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1024)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1001)

**Pull request merged**